packages/assets-controllers/src/TokenDetectionController.test.ts

```diff
@@ -1289,51 +1318,121 @@ async function withController<ReturnValue>(
   const controllerMessenger =
     messenger ?? new ControllerMessenger<AllowedActions, AllowedEvents>();
 
-  const mockGetNetworkConfigurationByNetworkClientId = jest
-    .fn<
-      ReturnType<NetworkController['getNetworkConfigurationByNetworkClientId']>,
-      Parameters<NetworkController['getNetworkConfigurationByNetworkClientId']>
-    >()
-    .mockImplementation((networkClientId: string) => {
-      return mockNetworkConfigurations[networkClientId];
-    });
+  const mockKeyringState = jest.fn<KeyringControllerState, []>();
+  controllerMessenger.registerActionHandler(
+    'KeyringController:getState',
+    mockKeyringState.mockReturnValue({
+      isUnlocked: true,
+    } as unknown as KeyringControllerState),
+  );
+  const mockGetNetworkConfigurationByNetworkClientId = jest.fn<
+    ReturnType<NetworkController['getNetworkConfigurationByNetworkClientId']>,
+    Parameters<NetworkController['getNetworkConfigurationByNetworkClientId']>
+  >();
   controllerMessenger.registerActionHandler(
     'NetworkController:getNetworkConfigurationByNetworkClientId',
-    mockGetNetworkConfigurationByNetworkClientId,
+    mockGetNetworkConfigurationByNetworkClientId.mockImplementation(
+      (networkClientId: string) => {
+        return mockNetworkConfigurations[networkClientId];
+      },
+    ),
   );
-  const mockTokenListGetState = jest
-    .fn<TokenListState, []>()
-    .mockReturnValue({ ...getDefaultTokenListState() });
+  const mockTokenListState = jest.fn<TokenListState, []>();
   controllerMessenger.registerActionHandler(
     'TokenListController:getState',
-    mockTokenListGetState,
+    mockTokenListState.mockReturnValue({ ...getDefaultTokenListState() }),
+  );
+  const mockPreferencesState = jest.fn<PreferencesState, []>();
+  controllerMessenger.registerActionHandler(
+    'PreferencesController:getState',
+    mockPreferencesState.mockReturnValue({
+      ...getDefaultPreferencesState(),
+    }),
   );
 
-  const preferencesStateChangeListeners: ((state: PreferencesState) => void)[] =
-    [];
   const controller = new TokenDetectionController({
     networkClientId: NetworkType.mainnet,
-    onPreferencesStateChange: (listener) => {
-      preferencesStateChangeListeners.push(listener);
-    },
     getBalancesInSingleCall: jest.fn(),
     addDetectedTokens: jest.fn(),
     getTokensState: jest.fn().mockReturnValue(getDefaultTokensState()),
-    getPreferencesState: jest.fn().mockReturnValue({
-      ...getDefaultPreferencesState(),
-      useTokenDetection: true,
-    }),
+    trackMetaMetricsEvent: jest.fn(),
     messenger: buildTokenDetectionControllerMessenger(controllerMessenger),
     ...options,
   });
   try {
     return await fn({
       controller,
-      mockTokenListGetState,
+      mockKeyringGetState: (state: KeyringControllerState) => {
+        controllerMessenger.unregisterActionHandler(
```

Comment:
I know I've asked this before in a different place, but why do these helpers have to unregister the action handler? I don't believe the tests for TokenDetectionController involve any other controller.

Reply:
They were unregistering the action handlers registered earlier in `withController`, but looks like there's no issue with directly using the jest mocks for this. Removed here:  https://github.com/MetaMask/core/pull/3775/commits/ec9869a861fcd3486613542fa9d31b66606091da